functions/Invoke-DbaDatabaseShrink.ps1

```diff
@@ -0,0 +1,279 @@
+			}
+			
+			if ($exclude)
+			{
+				$dbs = $dbs | Where-Object { $exclude -notcontains $_.Name }
+			}
+			
+			foreach ($db in $dbs)
+			{
+				Write-Message -Level Verbose -Message "Processing $db on $instance"
+				
+				if ($db.IsAccessible -eq $false)
+				{
+					Stop-Function -Message "The database $db on server $instance is not accessible. Skipping database." -Continue
+				}
+				
+				$startingsize = $db.Size
+				$spaceAvailableMB = $db.SpaceAvailable/1024
+				$spaceused = $startingsize - $spaceAvailableMB
+				$desiredSpaceAvailable = ($PercentFreeSpace * $spaceused)/100
+				
+				Write-Message -Level Verbose -Message "Starting Size (MB): $startingsize"
+				Write-Message -Level Verbose -Message "Starting Freespace (MB): $([int]$spaceavailableMB)"
+				Write-Message -Level Verbose -Message "Desired Freespace (MB): $([int]$desiredSpaceAvailable)"
+				
+				if (($db.SpaceAvailable/1024) -le $desiredSpaceAvailable)
+				{
+					Write-Message -Level Warning -Message "Space Available ($spaceavailableMB) is less than or equal to the desired outcome ($desiredSpaceAvailable)"
+				}
+				else
+				{
+					if ($Pscmdlet.ShouldProcess("$db on $instance", "Shrinking from $([int]$startingsize) MB to $([int]$desiredSpaceAvailable) MB"))
+					{
+						if ($db.Tables.Indexes.Name -and $server.VersionMajor -gt 8)
+						{
+							Write-Message -Level Verbose -Message "Getting average fragmentation"
+							$startingfrag = (Invoke-Sqlcmd2 -ServerInstance $instance -Credential $SqlCredential -Query $sql -Database $db.name | Select-Object -ExpandProperty avg_fragmentation_in_percent | Measure-Object -Average).Average
+						}
+						else
+						{
+							$startingfrag = $null
+						}
+						
+						$start = Get-Date
+						
+						try
+						{
+							$db.Shrink($PercentFreeSpace, $ShrinkMethod)
+							$db.Refresh()
+							$db.RecalculateSpaceUsage()
+							$success = $true
+							$notes = $null
+						}
+						catch
+						{
+							$success = $false
+							$notes = $_.Exception.InnerException
+						}
+						
+						$end = Get-Date
+						$dbsize = $db.Size
+						Write-Message -Level Verbose -Message "Final size: $([int]$dbsize) MB"
+						
+						if ($db.Tables.Indexes.Name -and $server.VersionMajor -gt 8)
+						{
+							Write-Message -Level Verbose -Message "Refreshing indexes and getting average fragmentation"
+							$endingdefrag = (Invoke-Sqlcmd2 -ServerInstance $instance -Credential $SqlCredential -Query $sql -Database $db.name | Select-Object -ExpandProperty avg_fragmentation_in_percent | Measure-Object -Average).Average
+						}
+						else
+						{
+							$endingdefrag = $null
+						}
+						
+						$timespan = New-TimeSpan -Start $start -End $end
+						$ts = [timespan]::fromseconds($timespan.TotalSeconds)
+						$elapsed = "{0:HH:mm:ss}" -f ([datetime]$ts.Ticks)
+					}
+				}
+				
+				#$db.TruncateLog()
+				
+				if ($Pscmdlet.ShouldProcess("$db on $instance", "Showing results"))
```

Comment:
sure this has to be wrapped in shouldprocess ? it doesn't do anything other than recalculating space usage...